### PR TITLE
Kc blanks attention

### PR DESCRIPTION
**Pull request by ciebson123**
# Description
New attention works like that:
Is causal 
Non-blanks can't attend to blanks
Blanks can attend to previous non-blanks and previous blanks in the same block

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Comment by ciebson123** [comment]
https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=9aa280ab-9ebb-42b1-8d9f-d6847b75ca60&dash=charts&compare=MwFgrGA0CMQ

**Comment by maciek-pioro** [comment]
@ciebson123 checklist plz

File: research/blanks/utils.py [hunk 1]
```diff
@@ -120,6 +120,28 @@ def can_fit_blanks(
...
+    blanks_start = get_first_blanks_in_series(is_blank)
+    blanks_last = get_last_blanks_in_series(is_blank)
+    tensors = []
+    for blanks, firsts, lasts in zip(is_blank, blanks_start, blanks_last):
+        mask = torch.tril(torch.ones((len(blanks), len(blanks)))) == 0
+        for first_ind, last_ind in zip(firsts.nonzero(), lasts.nonzero()):
+            mask[last_ind + 1 :, first_ind : last_ind + 1] = True
+        tensors.append(mask)
+    return torch.stack(tensors)
+
```
Comment (maciek-pioro) [inline research/blanks/utils.py:143]: Can we see how much slower that is than normal attention

File: research/blanks/data.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+class LLMExample(object):
+    input_ids: List[int]
+    target_ids: List[int]
+    should_calculate_loss: List[
+        int
+    ]  # e.g. in BERT loss is not calculated over non-masked tokens
+    attention_mask: np.ndarray
+
+
+class LLMBatch:
```
Comment (maciek-pioro) [inline research/blanks/data.py:18]: LLMBatch -> BlanxBatch

File: research/blanks/data.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+from typing import List
+
+import numpy as np
+import torch
+from attr import dataclass
+
+
+@dataclass
+class LLMExample(object):
```
Comment (maciek-pioro) [inline research/blanks/data.py:9]: LLMExample -> BlanxExample

File: research/blanks/model.py [hunk 4]
```diff
@@ -124,13 +125,26 @@ def get_ff_layer(args):
...
     attention_layer_fun = lambda: llm.Attention(
         dmodel=args.dmodel,
         heads=args.n_att_heads,
-        causal=True,
         dhead=args.dhead,
         flash=True,
+        causal=True,
         init_type=args.init_type,
         init_scale=args.init_scale,
     )
```
Comment (maciek-pioro) [inline research/blanks/model.py:150]: Let's unify this and add an `if` for either choosing BlankAttention or llm.Attention (see how they do that in conditional.

Comment (ciebson123) [reply research/blanks/model.py]: Done

File: research/blanks/data.py (hunk 3 above)
Comment (ciebson123) [reply research/blanks/data.py]: Done

File: research/blanks/data.py (hunk 2 above)
Comment (ciebson123) [reply research/blanks/data.py]: Done

**Review by maciek-pioro — APPROVED** [review]
In the future it might be wiser to move the blanx attention manager to trainer - current setup seems like it might cause problems with checkpointing / fsdp.

**Pull request merged**